server/src/session.rs

```diff
@@ -0,0 +1,120 @@
+    pub fn new(client: tower_lsp::Client) -> Self {
+        Self {
+            client,
+            configuration_sets: Mutex::new(Vec::new()),
+            root_uri: RwLock::new(None),
+            built_in_slice_path: RwLock::new(String::new()),
+        }
+    }
+
+    // Update the properties of the session from `InitializeParams`
+    pub async fn update_from_initialize_params(
+        &self,
+        params: tower_lsp::lsp_types::InitializeParams,
+    ) {
+        // This is the path to the built-in Slice files that are included with the extension. It should always
+        // be present.
+        let built_in_slice_path = params
+            .initialization_options
+            .and_then(|opts| opts.get("builtInSlicePath").cloned())
+            .and_then(|v| v.as_str().map(str::to_owned))
+            .expect("builtInSlicePath not found in initialization options");
+        *self.built_in_slice_path.write().await = built_in_slice_path.clone();
+
+        // Use the root_uri if it exists temporarily as we cannot access configuration until
+        // after initialization. Additionally, LSP may provide the windows path with escaping or a lowercase
+        // drive letter. To fix this, we convert the path to a URL and then back to a path.
+        let root_uri = params
+            .root_uri
+            .and_then(|uri| uri.to_file_path().ok())
+            .and_then(|path| Url::from_file_path(path).ok())
+            .expect("root_uri not found in initialization parameters");
+        *self.root_uri.write().await = Some(root_uri.clone());
+    }
+
+    // Update the stored configuration sets by fetching them from the client.
+    pub async fn fetch_configurations(&self) {
+        let built_in_path = &self.built_in_slice_path.read().await;
+        let root_uri_guard = self.root_uri.read().await;
+        let root_uri = (*root_uri_guard).clone().expect("root_uri not set");
+
+        let params = vec![ConfigurationItem {
+            scope_uri: None,
+            section: Some("slice.configurations".to_string()),
+        }];
+
+        // Fetch the configurations from the client and parse them.
+        let configurations = self
+            .client
+            .configuration(params)
+            .await
+            .ok()
+            .map(|response| {
+                let config_array = &response
+                    .iter()
+                    .filter_map(|config| config.as_array())
+                    .flatten()
+                    .cloned()
+                    .collect::<Vec<_>>();
+                ConfigurationSet::parse_configuration_sets(config_array, &root_uri, built_in_path)
+            })
+            .unwrap_or_default();
+
+        // Update the configuration sets
+        self.update_configurations(configurations).await;
+    }
+
+    // Update the configuration sets from the `DidChangeConfigurationParams` notification.
+    pub async fn update_configurations_from(&self, params: DidChangeConfigurationParams) {
```

Comment:
From what? Maybe just remove the from?

Reply:
There is already `update_configurations` which is private. I renamed it to `update_configurations_from_params`